File: services/optimizer/haversine.py
"""
Calcul de distance Haversine et matrice de distance corrigée.
Pas de dépendance tierce — stdlib Python uniquement (D-09).
"""

import math
from typing import Sequence

EARTH_RADIUS_KM = 6371.0


def haversine_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """
    Distance Haversine entre deux points géographiques, en kilomètres.
    Formule standard (rayon Terre 6371 km).
    """
    r = math.radians
    dlat = r(lat2 - lat1)
    dlng = r(lng2 - lng1)
    a = (
        math.sin(dlat / 2) ** 2
        + math.cos(r(lat1)) * math.cos(r(lat2)) * math.sin(dlng / 2) ** 2
    )
    return 2 * EARTH_RADIUS_KM * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def distance_matrix(
    coords: Sequence[tuple[float, float]],
    correction_factor: float,
) -> list[list[float]]:
    """
    Matrice NxN de distances Haversine × facteur_correction routier (DEC-056).
    coords[0] = dépôt fictif.
    Diagonale nulle, matrice symétrique.
    """
    n = len(coords)
    return [
        [
            haversine_km(coords[i][0], coords[i][1], coords[j][0], coords[j][1])
            * correction_factor
            for j in range(n)
        ]
        for i in range(n)
    ]

File: apps/web/src/lib/optimizer/haversine.ts
/**
 * Distance Haversine et matrice de distance corrigée pour le solveur local
 * (Phase 06.12 — heuristique TS native).
 *
 * Port direct de `apps/web/py/solver/haversine.py` (D-02). Aucune dépendance
 * tierce, stdlib seule.
 */

export const EARTH_RADIUS_KM = 6371;

/**
 * Distance Haversine entre deux points géographiques, en kilomètres.
 * Formule standard (rayon Terre 6371 km).
 */
export function haversineKm(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const toRad = (d: number): number => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return 2 * EARTH_RADIUS_KM * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

/**
 * Matrice NxN de distances Haversine × facteur de correction routier (DEC-056).
 * `coords[0]` = dépôt fictif. Diagonale nulle, matrice symétrique.
 */
export function distanceMatrix(
  coords: ReadonlyArray<readonly [number, number]>,
  correctionFactor: number,
): number[][] {
  const n = coords.length;
  const m: number[][] = [];
  for (let i = 0; i < n; i += 1) {
    const row: number[] = [];
    const ci = coords[i]!;
    for (let j = 0; j < n; j += 1) {
      const cj = coords[j]!;
      row.push(haversineKm(ci[0], ci[1], cj[0], cj[1]) * correctionFactor);
    }
    m.push(row);
  }
  return m;
}
